### ISSUE TITLE
Handling Imbalanced Class Instance in YOLOv8 ?

### ISSUE THREAD

**Issue by aimad1234** [ultralytics/ultralytics#6204]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

my question is : does YOLOv8 offer any configuration options that could alleviate the problem, or resolve it in the data preparation phase?

### Additional

I have 5 classes in my dataset and I'm faced with an imbalance in the number of instances distributed over five distinct classes, as shown in the figure above.
![Screenshot 2023-11-08 at 12 35 39](https://github.com/ultralytics/ultralytics/assets/66996795/bd8c3ab3-5447-4fa2-8cff-824a0d57306a)

**Comment by glenn-jocher** [comment]
@aimad1234 hello,

Thank you for reaching out with your question regarding handling imbalanced class instances in YOLOv8.

Class imbalance is indeed a common challenge in training object detector models when some classes are underrepresented in the dataset compared to others. YOLOv8 has internal mechanisms to mitigate this issue during training, though they are not directly exposed as configurable options.

One of the primary ways YOLOv8 tackles class imbalance is through the use of Focal Loss, which automatically gives more weight to harder, less frequent examples and less weight to easier, more common examples. This can help in rebalancing the learning focus, especially when dealing with imbalanced datasets.

Additionally, YOLOv8 uses data augmentation strategies during training that can help provide a more balanced view of different classes by artificially augmenting underrepresented classes more frequently. Augmentations such as random scaling, cropping, and flipping can increase the variability and effective sample size of minority classes in your dataset.

In the data preparation phase, you could manually augment your dataset for the underrepresented classes or gather more data for those classes. Alternatively, carefully curated oversampling of the minority class instances or synthesizing new instances using techniques such as SMOTE could be considered outside the scope of YOLOv8 functionalities.

Balancing class instances is often dataset-specific and requires some experimentation. The best approach depends on the degree of imbalance and the particularities of your dataset. Monitoring the training progress and validating the performance of each class can guide you on whether additional balance-focused interventions are necessary.

If you continue to experience issues or would like more personalized insights on addressing class imbalance during your YOLOv8 training, feel free to describe your situation further, and we can delve deeper into the specifics of your use case.

Wishing you success with your YOLOv8 model training!

**Comment by aimad1234** [comment]
@glenn-jocher 
It is very kind of you to give me this information, I will contact you again if I encounter any difficulties.

**Comment by glenn-jocher** [comment]
Hello @aimad1234,

You're most welcome! I'm glad to hear that you found the information helpful. Remember, tackling class imbalance can be quite nuanced, and it's all about finding the right approach that works for your specific dataset.

If you do run into any challenges or have more questions as you work with YOLOv8, don't hesitate to reach out. We're here to help you fine-tune your model and address any concerns that might arise during your training process. Your success is our priority, and we're committed to supporting our users through their machine learning journey.

Wishing you the best of luck with your YOLOv8 project! Happy training!

**Comment by glenn-jocher** [comment]
@tfp10 hello again,

Focal loss is an integral part of the loss function used by YOLOv8, designed to assist with class imbalance by default. This means when you train a model using the standard installation and procedures, the focal loss mechanism is already at work, helping to mitigate the imbalance by focusing on hard-to-classify examples.

There's no need for manual alterations to the source code or special configuration if you're using the version of YOLOv8 obtained through `pip`. The focal loss component is built-in and functioning as part of the training process automatically.

Should you wish to further adjust the balance or impact of focal loss beyond what's provided by default, that would involve a deeper dive into the source code. However, I suggest you first try training your model with the standard setup to see how it performs, as many users find it quite adequate without any need for customization.

If the results are not satisfying and you feel a need to adjust the focal loss, diving into the source code would indeed be the next step. However, I'd recommend this only for advanced users who are comfortable making changes and understand the implications on training stability and model performance.

Thank you for your kind words—supporting the community is a cornerstone of what we do. If you have further questions or need more assistance, please let us know.

Best regards to your endeavors with YOLOv8!

**Comment by tfp10** [comment]
Hi @glenn-jocher,

Thank you for your answer. 

When I checked the source code, there is indeed an implementation of Focal Loss. 
But looking specifically at the relevant implementation of the detection loss, I found in the class v8DetectionLoss: 
``
self.bce = nn.BCEWithLogitsLoss(reduction='none')
and later in the class: 

# loss[1] = self.varifocal_loss(pred_scores, target_scores, target_labels) / target_scores_sum  # VFL way
loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE
...
loss[0] *= self.hyp.box  # box gain
loss[1] *= self.hyp.cls  # cls gain
loss[2] *= self.hyp.dfl  # dfl gain

``

For my understandig, only normal BCE is applied instead an implemenation of a focal loss. 


Do I misunderstand something here?

Thank you in advance. 

Best regards

**Comment by glenn-jocher** [comment]
Hello @tfp10,

I appreciate your diligence in reviewing the source code and raising this point of clarification.

It appears you’ve stumbled upon a section of the code that seems to utilize the standard BCE (Binary Cross-Entropy) loss rather than the Focal Loss variant in the lines highlighted. This can be a bit confusing, as the mention of Focal Loss in our prior communications might suggest its application across the board.

In YOLOv8, while the core concept of Focal Loss is implemented, it may not be as a separate, easily identifiable function or parameter, especially if it’s been tightly integrated into the loss calculation workflow. This integration can involve adjusting the cross-entropy loss dynamically during training based on the confidence of the predictions, effectively mimicking the behavior of Focal Loss.

As the commented-out line indicates, there may be different loss formulations available within the codebase, which could be subject to testing or may reflect previous iterations of the design.

Understanding and modifying loss functions at this level indeed requires a deeper familiarity with the code. If you are looking to experiment with different loss configurations, it does require manual modification and certainly a thorough understanding of the implications each change might have on the model's performance.

If you feel the need to use a different loss function or adjust parameters associated with Focal Loss, you would need to modify the code accordingly. However, such changes should be made with caution and ideally tested thoroughly to ensure they produce the desired outcome.

As always, we're here to support your journey with YOLOv8, and your inquiries help improve not only your understanding but also the clarity of the information we provide to all users.

Best regards and I wish you success with your modifications and training!

**Comment by Y-T-G** [comment]
I have created a guide that doesn't require you to modify the loss function and is really straightforward:

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by Abhijeet241093** [comment]
Your guidline hard to understand. Thank you

**Comment by Abhijeet241093** [comment]
How to fine tune yolov8 model again with Focal Loss ?  @glenn-jocher @Y-T-G

**Comment by Abhijeet241093** [comment]
where is located hyp.yaml ?

**Comment by glenn-jocher** [comment]
The `hyp.yaml` file is not included in the default Ultralytics installation. For YOLOv8, you create a custom hyperparameters file in your project directory and reference it with the `hyp` argument during training (e.g., `model.train(hyp='hyp.yaml')`). You can start from the default hyperparameters shown in the [YOLOv8 training documentation](https://docs.ultralytics.com/modes/train/#arguments) and modify values as needed for your use case.

**Comment by Abhijeet241093** [comment]
Thank you for your guidance.

**Comment by glenn-jocher** [comment]
For guidance on creating a `hyp.yaml` file, start with the default hyperparameters from the [YOLOv8 training documentation](https://docs.ultralytics.com/modes/train/#arguments). Copy these values into a new `hyp.yaml` file in your project directory and modify them as needed. You can then reference this file in your training command with `model.train(hyp='hyp.yaml')`. Let us know if you encounter any specific challenges! 😊

**Comment by Abhijeet241093** [comment]
Thank you very much @glenn-jocher @Y-T-G for wonderful guidance.

**Comment by glenn-jocher** [comment]
You're very welcome, @Abhijeet241093! We're glad we could help. The Ultralytics team and the community are always here to support your efforts. Happy training!

**Comment by Abhijeet241093** [comment]
Thank you :)~

**Comment by glenn-jocher** [comment]
Hello @Abhijeet241093,

I noticed you're still investigating Focal Loss implementation for class imbalance. Looking at the reference documentation you're discussing, I can provide more technical details:

YOLOv8 includes both `VarifocalLoss` and `FocalLoss` implementations in the `ultralytics/utils/loss.py` file. The `VarifocalLoss` specifically addresses class imbalance as described in the paper (https://arxiv.org/abs/2008.13367).

To implement this in your model:

1. In the `v8DetectionLoss` class, locate where BCE is used:
   ```python
   loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE
   ```

2. Replace it with the VarifocalLoss implementation (uncomment the line):
   ```python
   loss[1] = self.varifocal_loss(pred_scores, target_scores, target_labels) / target_scores_sum  # VFL way
   ```

3. Set `use_vfl=True` in the class initialization

This is a code-level change rather than a simple configuration option, but it directly leverages the existing loss function implementations documented in the reference material.

**Comment by YousraAndoulla** [comment]
🚗 YOLOv8 – Vehicle Type Detection Issues (Misclassifications + Missed Detections)
Hi all 👋,

I'm training a 5-class vehicle type detection model with YOLOv8 using bounding boxes.
Classes: car, truck, motorcycle, bus, bicycle
📦 Dataset was annotated using Roboflow automatic labeling and exported in YOLOv8 format.

❗ Observations during validation:
Trucks, motorcycles, and buses are often misclassified as cars

Some cars are not detected at all

This happens inconsistently across validation batches

Cars dominate the training dataset. 
 How can I best handle class imbalance in YOLOv8?
Could this be purely due to dataset imbalance, or is the model confusing visual features between cars and larger vehicles?

What's the best way to analyze model bias (e.g., per-class precision/recall confusion)?

Is there any built-in support in YOLOv8 for using focal loss or class weights to reduce dominant-class bias?

Would manually augmenting rare classes (rotation, brightness, cropping) improve class separation?

Are there recommended post-processing strategies to reduce misclassification between visually similar classes (e.g., car vs. truck)?

Can I visualize intermediate outputs or attention maps to understand what features the model is focusing on?

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082
